providers/common/macs/cmac.c

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright 2018 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/opensslconf.h>
+#ifndef OPENSSL_NO_CMAC
+
+# include <openssl/core_numbers.h>
+# include <openssl/params.h>
+# include <openssl/evp.h>
+# include <openssl/cmac.h>
+
+# include "internal/provider_algs.h"
+
+/*
+ * Forward declaration of everything implemented here.  This is not strictly
+ * necessary for the compiler, but provides an assurance that the signatures
+ * of the functions in the dispatch table are correct.
+ */
+static OSSL_OP_mac_newctx_fn cmac_new;
+static OSSL_OP_mac_copyctx_fn cmac_copy;
+static OSSL_OP_mac_freectx_fn cmac_free;
+static OSSL_OP_mac_set_param_types_fn cmac_set_param_types;
+static OSSL_OP_mac_ctx_set_params_fn cmac_set_params;
+static OSSL_OP_mac_size_fn cmac_size;
+static OSSL_OP_mac_init_fn cmac_init;
+static OSSL_OP_mac_update_fn cmac_update;
+static OSSL_OP_mac_final_fn cmac_final;
+
+/* local CMAC pkey structure */
+
+/*
+ * TODO(3.0) solve the ENGINE problem
+ *
+ * For the moment, ENGINEs are simply ignored.  A tentative idea is to
```

Comment:
I don't think we can just ignore this problem. We shouldn't be introducing temporary code breaks. We should instead fix the underlying problem first.

Reply:
Like I said elsewhere, I've started working in ENGINE by name.  This comment will disappear.

Reply:
Comment removed, engines are now handled here.